src/cpp/session/modules/SessionObjectExplorer.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * SessionObjectExplorer.cpp
+ *
+ * Copyright (C) 2009-17 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#define R_INTERNAL_FUNCTIONS
+
+#include "SessionObjectExplorer.hpp"
+
+#include <boost/bind.hpp>
+#include <boost/foreach.hpp>
+
+#include <core/Algorithm.hpp>
+#include <core/Error.hpp>
+#include <core/Exec.hpp>
+
+#include <r/RExec.hpp>
+#include <r/RRoutines.hpp>
+#include <r/RSexp.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {      
+namespace explorer {
+
+namespace {
+
+const char * const kExplorerCacheDir = "explorer-cache";
+
+FilePath explorerCacheDir() 
+{
+   return module_context::sessionScratchPath()
+         .childPath(kExplorerCacheDir);
+}
+
+std::string explorerCacheDirSystem()
+{
+   return string_utils::utf8ToSystem(explorerCacheDir().absolutePath());
+}
+
+void removeOrphanedCacheItems()
```

Comment:
Will do! Initially, I intended to attach cache data items to the `.rs.CachedDataEnv`, but try as I might I couldn't convince RStudio not to delete my R data objects on suspend / resume / shutdown.